dlib/control/sarsa.h

```diff
@@ -0,0 +1,171 @@
+            double eps = 0.1,
+            bool v = false
+        ) : iterations(miters), verbose(v) {
+            set_learning_rate(lr);
+            set_discount(disc);
+            set_epsilon(eps);
+        }
+
+        double get_learning_rate(
+        ) const { return learning_rate; }
+
+        void set_learning_rate(
+            double value
+        )
+        {
+            DLIB_ASSERT(value >= 0. && value <= 1.,
+                "\t sarsa::set_learning_rate(value)"
+                "\n\t invalid inputs were given to this function"
+                "\n\t value: " << value
+            );
+            learning_rate = value;
+        }
+
+        double get_discount(
+        ) const { return discount; }
+
+        void set_discount(
+            double value
+        )
+        {
+            DLIB_ASSERT(value >= 0. && value <= 1.,
+                "\t sarsa::set_discount(value)"
+                "\n\t invalid inputs were given to this function"
+                "\n\t value: " << value
+            );
+            discount = value;
+        }
+
+        unsigned int get_iterations(
+        ) const { return iterations; }
+
+        void set_iterations(
+            unsigned int value
+        ) { iterations = value; }
+
+        double get_epsilon(
+        ) const { return epsilon; }
+
+        void set_epsilon(
+            double value
+        )
+        {
+            DLIB_ASSERT(value >= 0. && value <= 1.,
+                "\t sarsa::set_epsilon(value)"
+                "\n\t invalid inputs were given to this function"
+                "\n\t value: " << value
+            );
+            epsilon = value;
+        }
+
+        bool is_verbose(
+        ) const { return verbose; }
+
+        void be_verbose(
+        ) { verbose = true; }
+
+        void be_quiet(
+        ) { verbose = false; }
+
+        template <
+            typename policy_type,
+            typename prng_engine = std::default_random_engine
+            >
+        policy_type train_policy(
+            const policy_type &policy,
+            const prng_engine &gen = prng_engine()
+        ) const
+        {
+            typedef typename std::decay<decltype(policy.get_model())>::type::reward_type reward_type;
+
+            if(verbose)
+                std::cout << "Starting training..." << std::endl;
+
+            const auto &model = policy.get_model();
+            epsilon_policy<policy_type, prng_engine> eps_pol(epsilon, policy, gen);
+            auto& w = eps_pol.get_weights();
+
+            DLIB_ASSERT(weights.size() == model.states_size(),
+                "\t sarsa::train(weights)"
+                "\n\t invalid inputs were given to this function"
+                "\n\t weights.size: " << weights.size() <<
+                "\n\t features size: " << model.states_size()
+            );
+
+            reward_type total_reward = static_cast<reward_type>(0);
+            for(auto iter = 0u; iter < iterations; ++iter){
```

Comment:
Put { on a line by itself so the code style is consistent with all the other code in this file and other parts of dlib.